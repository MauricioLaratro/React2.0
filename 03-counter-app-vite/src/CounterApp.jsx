//TAREA
import PropTypes from 'prop-types';
import { useState } from 'react'

// Lo que hacemos aca es pasar la prop del componente padre que es value, con su valor predeterminado de 10, ese valor lo pasamos como valor inicial de nuestro useState del counter y lo que realmente renderizamos en el H2 es el counter. Es decir que si en el use estate cambiamos el valor inicial de counter, por otro que no sea "value" se renderizara ese nuevo valor y la prop value ya no se estara utilizando en el renderizado.
export const CounterApp = ({ value }) => {
    
    const [ counter, setCounter ] = useState( value );
    
    const handleAdd = () => {
        setCounter( counter + 1 )
    }

    const handleReduce = () => {
        setCounter( counter - 1 )
    }

    const handleReset = () => {
        setCounter( value )
    }

  return (
    <>
        <h1>
            CounterApp
        </h1>
        <h2>
            { counter }
        </h2>

        <button onClick={ handleAdd }> +1 </button>
        <button onClick={ handleReduce }> -1 </button>
        <button aria-label='btn-reset' onClick={ handleReset }> Reset </button>
    </>
  )
}


CounterApp.propTypes = {
    value: PropTypes.number.isRequired
}