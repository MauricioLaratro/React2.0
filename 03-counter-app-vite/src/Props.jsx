// Importamos la libreria PropTypes para validar las propiedades, asi evitamos errores y ademas es una buena practica
import PropTypes from 'prop-types';


export const Props = ({ title, subTitle, name }) => {
    return (
        <>
            <h1 data-testid='test-title'>
                { title }
            </h1>
            <p>
                { subTitle }
            </p>
            <p>
                { subTitle }
            </p>
            <p>
                { name }
            </p>
        </>
    )
}

// Validamos las props del componente prop de la siguiente manera. En donde la prop title estamos indicando que debe ser si o si un string y es requerido, es decir que en el caso que no se declado el string del title, devolvera un error en consola
Props.propTypes = {
    title: PropTypes.string.isRequired,
    subTitle: PropTypes.string,
    name: PropTypes.string,
};

Props.defaultProps = {
    title: 'No se definio un titutlo',
    subTitle: 'No se definio un subtitulo',
    name: 'No se definio un nombre'
}