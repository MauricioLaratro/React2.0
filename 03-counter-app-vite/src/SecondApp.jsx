
// Es un regla de react que si no se cambiara un elemento, o no se vera afectado por algun Hook, lo ubiquemos fuera de la function declarativa del componente, ya que evitamos re renderizados o reprocesos innecesarios. En el caso que el elemento se vea afectado por un useState por ejemplo, si deberia estar dentro de la function. (Al colocarlo fuera de la function no se coloca en el scope global, sigue estando encapsulado dentro del componente, nada mas que solo lo procesa en el primer renderizado).
// Simepre que no sea un Object, podemos utilizar nuestras variables en elementos dentro del componente para renderizarlos.
const appTitle = 'Second App!';

// Tambien podemos renderizar el resultado de una function, dentro del componente de React a renderizar
// A excepciones de las promesas y functiones async, ya que las promesas son en realidad Objetcs de JS por lo que no se pueden
// const getSaludo = ()=> 'Hola soy una function!'
// La siguiente function es lo mismo que la de arriba, pero utilizando los parametros para que sea variable, en vez de poner explicito el resultado en la function
const getSaludo = (a, b)=> a+b


export function SecondApp() {

    return (
        // Utiilzar <> vacio es una forma simplificada de utilizar Fragment de React, asi no tenemos que importarlo en el documento para poder utilizarlo.
        // Fragment sirve para tener 2 o mas elementos en el mismo nivel jerarquico, sin tener que envolverlos dentro de un <div>.
        <>
            <h2>
                { appTitle }
                <br/>
                { getSaludo('Hola ','Soy una function') }
            </h2>
            <p>
                Soy un P dentro del mismo componente utilizando Fragment de react
            </p>
        </>
    )
}


