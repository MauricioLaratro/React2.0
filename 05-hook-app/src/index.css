body {
    padding: 20px
}

button {
    margin-inline-end: 10px;
}

/* .todo-text{
    cursor: pointer;
    user-select: none;
} */

.progress-todo{
    background: #FFCA2C;
}

.todo-success{
    background: #79DCBC;
}

a {
    margin-right: 5px;
}