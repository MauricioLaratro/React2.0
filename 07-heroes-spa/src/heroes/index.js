// Archivo barril de heroes para obtimizar las importaciones

export * from './pages';
export * from './routes/HeroesRoutes';