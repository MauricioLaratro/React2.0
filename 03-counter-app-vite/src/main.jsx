import React from 'react';
import ReactDOM from 'react-dom/client';
import { App } from './FirstApp'
import { SecondApp } from './SecondApp';
import './index.css';
import { Props } from './Props';
import { CounterApp } from './CounterApp';

ReactDOM.createRoot( document.getElementById('root') ).render(
    <React.StrictMode>
        {/* <App /> */}
        {/* <SecondApp /> */}
        {/* <Props title='Titulo recibido a travÃ©s de las Props'/> */}
        {/* <Props/> */}
        <CounterApp value={ 20 }/>
    </React.StrictMode>
)