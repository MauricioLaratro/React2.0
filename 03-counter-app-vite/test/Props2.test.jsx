/** @jest-environment jsdom */
import { render, screen } from "@testing-library/react";
import { Props } from "../src/Props";

describe('Pruebas en Props', () => { 
  
  const title = 'Hola , soy goku'
  const subTitle = 'Hola , soy un subtitulo'

    test('Debe de hacer match con el snapshot', () => { 


      const { container } = render( <Props title={ title } /> )
      expect( container ).toMatchSnapshot()

     })

     test('Debe de mostrar el mensaje "Hola, soy Goku"', () => { 
      
      render( <Props title={ title } /> )
      expect( screen.getByText(title) ).toBeTruthy()

      // Screen es un metodo que importamos de React Testing Library, que sirve para tener una copia actualizada en todo momento, de lo que estamos renderizando, es decir que siempre debe ir debajo del render si lo utilizamos como en este ejemplo. A diferencia del snapshot, screen se actualizara cuando haya cambios en el DOM por ejemplo cuando simulemos clicks que afectan al DOM de alguna manera. Es una manera mas dinamica y flexible de testear los componentes.
      })


      test('Debe de mostrar el titulo en un H1', () => { 
      
        render( <Props title={ title } /> )
        expect( screen.getByRole('heading', { level: 1 }).innerHTML ).toContain( title )
  
        // con getByRole lo que hacemos es obtener el elemento mediante un tag HTML, en este caso le indicamos que es un H1 y que ese H1 debe contener "toContain" el title que almacena como valor nuestra constante title
        })


      test('Debe de mostrar el subTitulo enviado por props', () => { 
      
        render( <Props title={ title } subTitle={ subTitle } /> )  
        expect( screen.getAllByText(subTitle).length ).toBe(2)


      })

 })