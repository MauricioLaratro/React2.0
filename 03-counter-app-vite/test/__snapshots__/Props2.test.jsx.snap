// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Props Debe de hacer match con el snapshot 1`] = `
<div>
  <h1
    data-testid="test-title"
  >
    Hola , soy goku
  </h1>
  <p>
    No se definio un subtitulo
  </p>
  <p>
    No se definio un subtitulo
  </p>
  <p>
    No se definio un nombre
  </p>
</div>
`;
